diff -U 3 -Nr '--exclude=.*.sw[a-z]' '--exclude=bugfixes.patch' '--exclude=.git*' nethack-3.4.3/bugfixes.txt nethack-3.4.3-bugfix/bugfixes.txt
--- nethack-3.4.3/bugfixes.txt	1969-12-31 19:00:00.000000000 -0500
+++ nethack-3.4.3-bugfix/bugfixes.txt	2013-03-31 21:37:38.790166200 -0400
@@ -0,0 +1,10 @@
+List of bugs fixed in this repository:
+
+* C343-248:  You can't put candles in a candelabrum while underwater.
+             Author:  Sgeo
+             Source:  http://bilious.alt.org/?257
+
+* C343-324:  Cutting a long worm in two will crash the game if the cut takes
+             the worm to 1 HP or if long worms had become extinct.
+             Author:  Patric Mueller
+             Source:  http://bilious.alt.org/?332
diff -U 3 -Nr '--exclude=.*.sw[a-z]' '--exclude=bugfixes.patch' '--exclude=.git*' nethack-3.4.3/README nethack-3.4.3-bugfix/README
--- nethack-3.4.3/README	2003-12-07 18:39:13.000000000 -0500
+++ nethack-3.4.3-bugfix/README	2013-03-31 21:01:18.387440642 -0400
@@ -1,3 +1,19 @@
+This repository consists of the original vanilla NetHack 3.4.3 sources, plus
+fixes for bugs discovered and reported by the community -- primarily on
+NetHackWiki (http://nethackwiki.com/), on Bilious (http://bilious.alt.org/) and
+on rec.games.roguelike.nethack.  The aim is to provide "one stop shopping" for
+fixes for known bugs; Bilious, SourceForge, GitHub and the like are the places
+for new features.
+
+Priority is given to bugs listed by the DevTeam.
+
+The file "bugfixes.patch" contains the diff between the unmodified sources and
+this repository.  The file "bugfixes.txt" contains the list of bugs fixed.
+
+The original NetHack README follows:
+
+==============================================================================
+
                 NetHack 3.4.3 -- General information
 
 NetHack 3.4 is an enhancement to the dungeon exploration game NetHack.
diff -U 3 -Nr '--exclude=.*.sw[a-z]' '--exclude=bugfixes.patch' '--exclude=.git*' nethack-3.4.3/src/apply.c nethack-3.4.3-bugfix/src/apply.c
--- nethack-3.4.3/src/apply.c	2003-12-07 18:39:13.000000000 -0500
+++ nethack-3.4.3-bugfix/src/apply.c	2013-03-31 21:30:24.411403518 -0400
@@ -922,13 +922,13 @@
 		You(no_elbow_room);
 		return;
 	}
-	if(Underwater) {
-		pline("Sorry, fire and water don't mix.");
-		return;
-	}
 
 	otmp = carrying(CANDELABRUM_OF_INVOCATION);
 	if(!otmp || otmp->spe == 7) {
+		if(Underwater) {
+		    pline("Sorry, fire and water don't mix.");
+		    return;
+		}
 		use_lamp(obj);
 		return;
 	}
@@ -938,6 +938,10 @@
 		safe_qbuf(qbuf, sizeof(" to ?"), the(xname(otmp)),
 			the(simple_typename(otmp->otyp)), "it"));
 	if(yn(qbuf) == 'n') {
+		if(Underwater) {
+		    pline("Sorry, fire and water don't mix.");
+		    return;
+		}
 		if (!obj->lamplit)
 		    You("try to light %s...", the(xname(obj)));
 		use_lamp(obj);
diff -U 3 -Nr '--exclude=.*.sw[a-z]' '--exclude=bugfixes.patch' '--exclude=.git*' nethack-3.4.3/src/worm.c nethack-3.4.3-bugfix/src/worm.c
--- nethack-3.4.3/src/worm.c	2003-12-07 18:39:13.000000000 -0500
+++ nethack-3.4.3-bugfix/src/worm.c	2013-03-31 21:35:39.988686294 -0400
@@ -303,6 +303,23 @@
 	    (void) mattacku(worm);
 }
 
+/*  cutoff()
+ *
+ *  Remove the tail of a worm and adjust the hp of the worm.
+ */
+void
+cutoff(worm, tail)
+    struct monst *worm;
+    struct wseg *tail;
+{
+	if (flags.mon_moving)
+	    pline("Part of the tail of %s is cut off.", mon_nam(worm));
+	else
+	    You("cut part of the tail off of %s.", mon_nam(worm));
+	toss_wsegs(tail, TRUE);
+	if (worm->mhp > 1) worm->mhp /= 2;
+}
+
 /*  cutworm()
  *
  *  Check for mon->wormno before calling this function!
@@ -368,17 +385,15 @@
 
     /* Sometimes the tail end dies. */
     if (rn2(3) || !(new_wnum = get_wormno())) {
-	if (flags.mon_moving)
-	    pline("Part of the tail of %s is cut off.", mon_nam(worm));
-	else
-	    You("cut part of the tail off of %s.", mon_nam(worm));
-	toss_wsegs(new_tail, TRUE);
-	if (worm->mhp > 1) worm->mhp /= 2;
+	cutoff(worm, new_tail);
 	return;
     }
 
     remove_monster(x, y);		/* clone_mon puts new head here */
-    new_worm = clone_mon(worm, x, y);
+    if (!(new_worm = clone_mon(worm, x, y))) {
+	cutoff(worm, new_tail);
+	return;
+    }
     new_worm->wormno = new_wnum;	/* affix new worm number */
 
     /* Devalue the monster level of both halves of the worm. */
